R/cv_varsel.R

```diff
@@ -765,59 +806,53 @@ loo_varsel <- function(refmodel, method, nterms_max, ndraws,
-        # keeping in mind that we have `r_eff = 1` for all observations here).
-        S_for_psis_eval <- nrow(log_lik_ref)
-        no_psis_eval <- ceiling(min(0.2 * S_for_psis_eval,
-                                    3 * sqrt(S_for_psis_eval))) < 5
-        if (no_psis_eval) {
-          if (getOption("projpred.warn_psis", TRUE)) {
-            warning(
-              "Using standard importance sampling (SIS), as the number of ",
-              "draws or clusters is too small for PSIS. For improved ",
-              "accuracy, increase the number of draws or clusters, or use ",
-              "K-fold-CV."
-            )
-          }
-          # Use loo::sis().
-          # In principle, we could rely on loo::psis() here (because in such a
-          # case, it would internally switch to SIS automatically), but using
-          # loo::sis() explicitly is safer because if the loo package changes
-          # its decision rule, we would get a mismatch between our customized
-          # warning here and the IS method used by loo. See also loo issue
-          # stan-dev/loo#227.
-          importance_sampling_nm <- "sis"
-        } else {
-          # Use loo::psis().
-          # Usually, we have a small number of projected draws here (400 by
-          # default), which means that the 'loo' package will automatically
-          # perform the regularization from Vehtari et al. (2024,
-          # <https://jmlr.org/papers/v25/19-556.html>, appendix G).
-          importance_sampling_nm <- "psis"
-        }
-      } else {
+      # Take into account that clustered draws usually have different weights:
+      lw_sub <- log_lik_ref + log(refdist_eval$wdraws_prj)
+      # This re-weighting requires a re-normalization (as.array() is applied to
+      # have stricter consistency checks, see `?sweep`):
+      lw_sub <- sweep(lw_sub, 2, as.array(apply(lw_sub, 2, log_sum_exp)))
+      # Internally, loo::psis() doesn't perform the Pareto smoothing if the
+      # number of draws is small (as indicated by object `no_psis_eval`, see
+      # below). In projpred, this can occur, e.g., if users request a number
+      # of projected draws (for performance evaluation, either after
+      # clustering or thinning the reference model's posterior draws) that is
+      # much smaller than the default of 400. In order to throw a customized
+      # warning message (and to avoid the calculation of Pareto k-values, see
+      # loo issue stan-dev/loo#227), object `no_psis_eval` indicates whether
+      # loo::psis() would perform the Pareto smoothing or not (for the
+      # decision rule, see loo:::n_pareto() and loo:::enough_tail_samples(),
+      # keeping in mind that we have `r_eff = 1` for all observations here).
+      S_for_psis_eval <- nrow(log_lik_ref)
+      no_psis_eval <- ceiling(min(0.2 * S_for_psis_eval,
+                                  3 * sqrt(S_for_psis_eval))) < 5
+      if (no_psis_eval) {
```

Comment:
As far as I understand the changes here (and in the following code blocks; I just cannot comment on multiple code blocks at once), they imply not to use SIS anymore if the "projected draws" (i.e., the draws resulting from the clustering or thinning) have nonconstant weights (which should only occur in case of clustering because I think that projpred requires the original draws to have constant weights, so thinned draws would also have constant weights). Is that change on purpose? If yes, can we move it out into a separate PR to have as few changes as possible in the test snapshots?

Reply:
The relevant commit seems to be 1fa7fcd.

Reply:
I think the changes (now reverted, see https://github.com/stan-dev/projpred/pull/496#issuecomment-2417980645) introduce at least one bug: `lw_sub` was not used anymore after its initial definitions from lines
```r
lw_sub <- log_lik_ref + log(refdist_eval$wdraws_prj)
```
and
```r
lw_sub <- sweep(lw_sub, 2, as.array(apply(lw_sub, 2, log_sum_exp)))
```
(these definitions were later overwritten). Furthermore, shouldn't it be
```r
lw_sub <- -log_lik_ref + log(refdist_eval$wdraws_prj)
```
(i.e., with the negative `log_lik_ref` values) in the first definition?

I guess the idea behind all these SIS / PSIS changes was that PSIS is supposedly able to deal with draws that have nonconstant weights by pre-multiplying the SIS weights (which would usually be used as input for `loo::psis()`) with the nonconstant draw weights and then to use those modified SIS weights as input for `loo::psis()`. Is that correct? If yes, the bug(s) would need to be fixed and then I would propose to create a separate PR for this (see [below](https://github.com/stan-dev/projpred/pull/496#issuecomment-2417980645) for why I had to revert the subsampling-unrelated changes anyway).

Reply:
Marking this conversation as "resolved" because this problem has been circumvented (for this PR) by moving the SIS / PSIS changes (and some other changes) to branch [`misc_from_fix-subsampling`](https://github.com/stan-dev/projpred/tree/misc_from_fix-subsampling). In the PR for that `misc_from_fix-subsampling` branch, this needs to be taken into account.